Add reuse_wrapping_token functionality.

melbit-michaelw opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-vault/blob/master/.github/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #976 

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/approle_auth_backend_role_secret_id: Add support for re-using a wrapped secret id (#976)
```

Output from acceptance testing:

```
$ make testacc TESTARGS="--run TestAccAppRoleAuthBackendRoleSecretID"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./...) -v --run TestAccAppRoleAuthBackendRoleSecretID -timeout 120m
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_basic
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_basic (0.22s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped (0.20s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_reuse
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped_reuse (0.20s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace
    util.go:112: TF_ACC_ENTERPRISE is not set, test is applicable only for Enterprise version of Vault
--- SKIP: TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace (0.00s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_full
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_full (0.20s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     2.038s
```

NOTE - This functionality relies on Vault updates in https://github.com/hashicorp/vault/pull/12425 

I'm not sure if I'm supposed to wait for that code to be merged/rejected before raising this PR, but I figured this way I can at least get some potential feedback and address any issues in this PR so it's ready to go when/if that other code is available.

melbit-michaelw commented:
The code that this relies on has been released in Vault 1.9.0, so I believe this is ready to be reviewed.

melbit-michaelw commented:
Any chance of someone looking at this any time soon ?

benashz commented:
> Any chance of someone looking at this any time soon ?

@melbit-michaelw thanks for the PR, we will take a look at this week.

benashz commented on vault/resource_approle_auth_backend_role_secret_id_test.go:
```diff
@@ -57,6 +58,30 @@ func TestAccAppRoleAuthBackendRoleSecretID_wrapped(t *testing.T) {
+	role := acctest.RandomWithPrefix("test-role")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { util.TestAccPreCheck(t) },
```
`TestAccPreCheck()` was moved to the `testutil` package.

```suggestion
		PreCheck:     func() { testutil.TestAccPreCheck(t) },
```

benashz commented:
@melbit-michaelw would you mind rebasing this PR off of `main`.  There was some refactoring done on that branch that will result in some semantic merge errors.

See my comment https://github.com/hashicorp/terraform-provider-vault/pull/1166#discussion_r800842326. 

Thanks,

Ben

melbit-michaelw commented:
Hi @benashz,

I've rebased onto main, and fixed up the package for the TestAccePreCheck function. 

Updated output from acceptance tests:
```
$ TESTARGS="--run AppRoleAuthBackendRole" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -v --run AppRoleAuthBackendRole -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
?       github.com/hashicorp/terraform-provider-vault/testutil  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
=== RUN   TestAccAppRoleAuthBackendRoleID_basic
--- PASS: TestAccAppRoleAuthBackendRoleID_basic (7.05s)
=== RUN   TestAccAppRoleAuthBackendRoleID_customID
--- PASS: TestAccAppRoleAuthBackendRoleID_customID (5.46s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_basic
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_basic (3.18s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped (3.06s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_reuse
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped_reuse (3.34s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace
    testutil.go:57: "TF_ACC_ENTERPRISE" must be set
--- SKIP: TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace (0.00s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_full
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_full (3.01s)
=== RUN   TestAccAppRoleAuthBackendRole_import
--- PASS: TestAccAppRoleAuthBackendRole_import (3.62s)
=== RUN   TestAccAppRoleAuthBackendRole_basic
--- PASS: TestAccAppRoleAuthBackendRole_basic (3.00s)
=== RUN   TestAccAppRoleAuthBackendRole_update
--- PASS: TestAccAppRoleAuthBackendRole_update (5.24s)
=== RUN   TestAccAppRoleAuthBackendRole_full
--- PASS: TestAccAppRoleAuthBackendRole_full (2.87s)
=== RUN   TestAccAppRoleAuthBackendRole_fullUpdate
--- PASS: TestAccAppRoleAuthBackendRole_fullUpdate (7.61s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     47.973s
```

benashz commented on vault/resource_approle_auth_backend_role_secret_id.go:
```diff
@@ -79,6 +79,13 @@ func approleAuthBackendRoleSecretIDResource(name string) *schema.Resource {
 				},
 			},
 
+			"reuse_wrapping_token": {
```
I feel like `reuse_wrapping_token` might be a bit confusing for an operator, in that it somehow might relate to the way that the provider will be accessing vault,  but what it seems to be is a directive for specifying which accessor is provided in `wrapping_accessor`. Maybe we could rename it to `with_wrapped_accessor`

An alternative approach would be to rename it to `wrapping_accessor_type` with `schema.TypeString`, that takes either `default`, or `wrapped` as possible values?

melbit-michaelw replied:
reuse_wrapping_token was originally suggested by @jasonodonnell  ( https://github.com/hashicorp/terraform-provider-vault/issues/976#issuecomment-862493014 ). I've got no particular attachment to it, so I'll change it to with_wrapped_accessor as suggested. 

Personally, I prefer boolean over a 2 possible value string, but if you really want it to be the second option you suggested, let me know and I'll see if I can make that work.

benashz replied:
Seems good!

benashz commented on vault/resource_approle_auth_backend_role_secret_id.go:
```diff
@@ -147,6 +154,8 @@ func approleAuthBackendRoleSecretIDCreate(d *schema.ResourceData, meta interface
 	} else {
 		data["metadata"] = ""
 	}
+	var reuseWrappingToken bool
+	reuseWrappingToken = d.Get("reuse_wrapping_token").(bool)
```
I think we can safely omit the boolean declaration, replacing it with:
```suggestion
	reuseWrappingToken := d.Get("reuse_wrapping_token").(bool)
```

benashz commented on vault/resource_approle_auth_backend_role_secret_id_test.go:
```diff
@@ -58,6 +59,30 @@ func TestAccAppRoleAuthBackendRoleSecretID_wrapped(t *testing.T) {
+					resource.TestCheckResourceAttrSet(secretIDResource, "wrapping_accessor"),
+					resource.TestCheckResourceAttrSet(secretIDResource, "wrapping_token"),
+					resource.TestCheckResourceAttrSet(secretIDResource, "accessor"),
+					resource.TestMatchResourceAttr(secretIDResource, "accessor", regexp.MustCompile("^[{]?[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}[}]?$")),
```
nit: I think we can probably simplify the regex pattern a bit by using the the `[[:xdigit:]]` hex digit character class like:
```suggestion
					resource.TestMatchResourceAttr(secretIDResource, "accessor",
						regexp.MustCompile("^[[:xdigit:]]{8}-([[:xdigit:]]{4}-){3}[[:xdigit:]]{12}$")),
```

benashz commented on vault/resource_approle_auth_backend_role_secret_id_test.go:
```diff
@@ -58,6 +59,30 @@ func TestAccAppRoleAuthBackendRoleSecretID_wrapped(t *testing.T) {
+			{
+				Config: testAccAppRoleAuthBackendRoleSecretIDConfig_wrapped_reuse(backend, role),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(secretIDResource, "backend", backend),
```
Would probably be a good idea to check the value of the field being added.

melbit-michaelw replied:
For the code snippet you've linked this to, I believe from the description that TestCheckResourceAttr compares the data in the secretIDResource with the key 'backend' against the value of backend.. 

i.e. I believe this is already being tested.

For the other attributes, 'wrapping_accessor', 'wrapping_token' and 'accessor', I'm not sure how to check the value since it's random generated by Vault.

Have I misunderstood something here ?

benashz commented on vault/resource_approle_auth_backend_role_secret_id.go:
```diff
@@ -330,8 +345,8 @@ func approleAuthBackendRoleSecretIDExists(d *schema.ResourceData, meta interface
-func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool) string {
-	if wrapped {
+func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool, reuse_wrapping_token bool) string {
+	if wrapped && !reuse_wrapping_token {
```
I wonder if it might make sense to include a special path prefix in the case that `reuseWrappingToken` is true?

melbit-michaelw replied:
I'm not sure what you are intending here.. Could you please elaborate on what you mean, if this is required ?

benashz replied:
I think it should be okay as-is.

benashz commented on vault/resource_approle_auth_backend_role_secret_id.go:
```diff
@@ -330,8 +345,8 @@ func approleAuthBackendRoleSecretIDExists(d *schema.ResourceData, meta interface
 
-func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool) string {
-	if wrapped {
+func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool, reuse_wrapping_token bool) string {
```
Go variable/argument name should be camel cased.
```suggestion
func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool, reuseWrappingToken bool) string {
```

benashz commented on vault/resource_approle_auth_backend_role_secret_id.go:
```diff
@@ -330,8 +345,8 @@ func approleAuthBackendRoleSecretIDExists(d *schema.ResourceData, meta interface
-func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool) string {
-	if wrapped {
+func approleAuthBackendRoleSecretIDID(backend, role, accessor string, wrapped bool, reuse_wrapping_token bool) string {
+	if wrapped && !reuse_wrapping_token {
```
```suggestion
	if wrapped && !reuseWrappingToken {
```

benashz reviewed (commented):
Looking pretty good. I have a few comments/suggestions to look at.

melbit-michaelw commented:
Updated output from acceptance tests:

```
$ TESTARGS="--run AppRoleAuthBackendRole" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -v --run AppRoleAuthBackendRole -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
?       github.com/hashicorp/terraform-provider-vault/testutil  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
=== RUN   TestAccAppRoleAuthBackendRoleID_basic
--- PASS: TestAccAppRoleAuthBackendRoleID_basic (2.44s)
=== RUN   TestAccAppRoleAuthBackendRoleID_customID
--- PASS: TestAccAppRoleAuthBackendRoleID_customID (1.84s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_basic
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_basic (1.23s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped (1.03s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_withWrappedAccessor
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_wrapped_withWrappedAccessor (0.98s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace
    testutil.go:57: "TF_ACC_ENTERPRISE" must be set
--- SKIP: TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace (0.00s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace_withWrappedAccessor
    testutil.go:57: "TF_ACC_ENTERPRISE" must be set
--- SKIP: TestAccAppRoleAuthBackendRoleSecretID_wrapped_namespace_withWrappedAccessor (0.00s)
=== RUN   TestAccAppRoleAuthBackendRoleSecretID_full
--- PASS: TestAccAppRoleAuthBackendRoleSecretID_full (1.00s)
=== RUN   TestAccAppRoleAuthBackendRole_import
--- PASS: TestAccAppRoleAuthBackendRole_import (1.23s)
=== RUN   TestAccAppRoleAuthBackendRole_basic
--- PASS: TestAccAppRoleAuthBackendRole_basic (0.95s)
=== RUN   TestAccAppRoleAuthBackendRole_update
--- PASS: TestAccAppRoleAuthBackendRole_update (1.65s)
=== RUN   TestAccAppRoleAuthBackendRole_full
--- PASS: TestAccAppRoleAuthBackendRole_full (0.98s)
=== RUN   TestAccAppRoleAuthBackendRole_fullUpdate
--- PASS: TestAccAppRoleAuthBackendRole_fullUpdate (2.40s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     16.076s
```

joetito1 commented:
👍

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-vault?pullRequest=1166)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-vault?pullRequest=1166)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-vault?pullRequest=1166).</sub>

benashz reviewed (approved):
LGTM!

Thank you for your contribution to HashiCorp.

melbit-michaelw commented:
@benashz 

Hey, could you please confirm that this made it into the 3.4.0 release ? Just confirming, since the changelog for that release doesn't appear to mention it. 

Also the documentation hasn't been updated, was I supposed to do that as part of this PR ?

nvx commented:
> Hey, could you please confirm that this made it into the 3.4.0 release ? Just confirming, since the changelog for that release doesn't appear to mention it.

https://github.com/hashicorp/terraform-provider-vault/commit/c324dc8dee48a40cff933c39864df734e7010b53 looks to be included in the v3.4.0 tag at least.

benashz commented:
@melbit-michaelw @nvx yes it made it in. Looks like we missed the change log entry. We will correct that on the next patch release.
